agent/php_user_instrument_wraprec_hashmap.c

```diff
@@ -0,0 +1,485 @@
+
+  if (hashmap->buckets[hash_key]) {
+    hashmap->buckets[hash_key]->prev = bucket;
+  }
+
+  hashmap->buckets[hash_key] = bucket;
+  ++hashmap->elements;
+
+  return bucket->scoped_funcs_ht;
+}
+
+static nr_func_hashmap_t* nr_scope_hashmap_lookup_internal(nr_scope_hashmap_t* hashmap, nr_scope_hashmap_key_t* key) {
+  size_t hash;
+  nr_scope_bucket_t* bucket;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_scope_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_scope_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->scoped_funcs_ht;
+  }
+
+  return NULL;
+}
+
+static nr_func_hashmap_t* nr_scope_hashmap_update_internal(nr_scope_hashmap_t* hashmap, nr_scope_hashmap_key_t* key) {
+  size_t hash;
+  nr_scope_bucket_t* bucket;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_scope_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_scope_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->scoped_funcs_ht;
+  }
+
+  return nr_scope_hashmap_add_internal(hashmap, hash, key);
+}
+
+static void nr_scope_hashmap_destroy_bucket_internal(nr_scope_bucket_t** bucket_ptr) {
+  nr_scope_bucket_t* bucket = *bucket_ptr;
+  nr_free(bucket->key->name);
+  nr_free(bucket->key);
+  nr_func_hashmap_destroy_internal(&bucket->scoped_funcs_ht);
+  nr_realfree((void**)bucket_ptr);
+}
+
+static void nr_scope_hashmap_destroy_internal(nr_scope_hashmap_t** hashmap_ptr) {
+  size_t count;
+  nr_scope_hashmap_t* hashmap;
+  size_t i;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return;
+  }
+  hashmap = *hashmap_ptr;
+
+  count = (size_t)(1 << hashmap->log2_num_buckets);
+  for (i = 0; i < count; i++) {
+    nr_scope_bucket_t* bucket = hashmap->buckets[i];
+
+    while (bucket) {
+      nr_scope_bucket_t* next = bucket->next;
+
+      nr_scope_hashmap_destroy_bucket_internal(&bucket);
+
+      bucket = next;
+    }
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+}
+
+nr_func_hashmap_t* global_funcs_ht = NULL;
+nr_scope_hashmap_t* scope_ht = NULL;
+
+static void nr_php_user_instrument_wraprec_hashmap_name2keys(
+  nr_func_hashmap_key_t* func,
+  nr_scope_hashmap_key_t* scope,
+  const char* full_name,
+  int full_name_len) {
+
+  if (nrunlikely(NULL == full_name)) {
+    return;
+  }
+  if (nrunlikely(full_name_len <= 0)) {
+    return;
+  }
+
+  func->name = full_name;
+  func->name_len = full_name_len;
+  func->name_hash = 0;
+  scope->name = NULL;
+  scope->name_len = 0;
+  scope->name_hash = 0;
+
+  /* If scope::method, then break into two strings */
+  for (int i = 0; i < full_name_len; i++) {
+    if ((':' == full_name[i]) && (':' == full_name[i + 1])) {
+      func->is_method = true;
+      scope->name = full_name;
+      scope->name_len = i;
+      func->name = full_name + i + 2;
+      func->name_len = full_name_len - i - 2;
```

Comment:
What if '::' was last two chars in string and `func->name_len` was 0 how is this handled by rest of code?

Reply:
This code is almost a verbatim copy of [this code](https://github.com/newrelic/newrelic-php-agent/blob/refactor/optimize-wraprec-lookup/agent/php_user_instrument.c#L338-L358). Should this concern be addressed in both places?

Reply:
I'll add a check for last character being `:` to `nr_php_add_custom_tracer_named`.

Reply:
Input validation added in adccd7f9c6f9de98970c52cf9d1a9c16d894cb73.

Reply:
The change looks good to me, just had request for two additional comments to clarify the func is no longer validating and expects a prevalidated value.
https://github.com/newrelic/newrelic-php-agent/pull/1042#discussion_r2067357743
https://github.com/newrelic/newrelic-php-agent/pull/1042#discussion_r2067364238